all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots for that task type at all
+  times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For example,
+  in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
+
+  /// Tries to immediately reserve a slot, returning true if a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error>;
+
+  /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
+  /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
+  /// then the slot can actually be used to work on the task.
+  /// 
+  /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
+  /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
+  fn mark_slot_used(&self, info: &Self::SlotKind::Info);
+
+  /// Frees a slot.
+  fn release_slot(&self, info: &Self::SlotKind::Info);
+  
+  /// If this implementation knows how many slots are available at any moment, it should return that here.
+  fn available_slots(&self) -> Option<usize>;
```

Comment:
I don't think this is necessary on this interface. That an implementation is has this information is not shouldn't matter to the system (it can emit its own available slot metric, we don't need to abstract it).